### PR TITLE
Optimized ExistsEdge by using a HashCode based HashSet<Edge> collecti…

### PR DESCRIPTION
**Pull request by uffelauesen**
### Issues

_This pull request fixes issue #714. Originally reported as part of issue #662. Relates to PR #663._  
### Description

_In BindingGraph you have OutgoingEdges implemented as a List<Edge>. When new entities are loaded into a DSC<> a check is performed agains the OutgoingEdges that uses a heavy Ling over the List<> collection. I have changed this collection to be a HashCode based HasSet<> - this speeds up performance dramatically when many items are loaded._
### Checklist (Uncheck if it is not completed)
- [  ] Test cases added
- [  ] Build and test with one-click build and test script passed
### Additional work necessary

_If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue._

…on for OutgoingEdges. Changed from a Lists<Edge> that used heavy Linq for detection of existance.

### CODE REVIEW

File: WCFDataService/Client/System/Data/Services/Client/Binding/BindingGraph.cs [hunk 1]
```diff
@@ -1146,6 +1167,46 @@ public bool Equals(Edge other)
...
+            /// <returns>The computed HashCode</returns>
+            public int GetHashCode(Edge obj)
+            {
+                if (obj == null)
+                    return 0;
+
+                unchecked
+                {
+                    int hash = 17;
+                    hash = hash * 23 + (obj.Source == null ? 0 : obj.Source.GetHashCode());
```
Comment (uffelauesen) [inline WCFDataService/Client/System/Data/Services/Client/Binding/BindingGraph.cs]: They are just some prime numbers. Basically we want to ensure that (x, y, s) does not return the same hash code as (y, x, s).

File: WCFDataService/Client/System/Data/Services/Client/Binding/BindingGraph.cs [hunk 2]
```diff
@@ -840,8 +840,8 @@ public void RemoveEdge(object source, object target, string label)
             /// <returns>true if an edge exists between source and target with given label, false otherwise</returns>
             public bool ExistsEdge(object source, object target, string label)
             {
-                Edge e = new Edge { Source = this.vertices[source], Target = this.vertices[target], Label = label };
-                return this.vertices[source].OutgoingEdges.Any(r => r.Equals(e));
+                Edge e = new Edge(this.vertices[source], this.vertices[target], label);
+                return this.vertices[source].OutgoingEdges.Contains(e);
```
Comment (uffelauesen) [inline WCFDataService/Client/System/Data/Services/Client/Binding/BindingGraph.cs]: EdgeComparer GetHashCode is called automatically by the HashSet when an edge is added to the set. The returned hash code is used internally by the HashSet to check for equality - and thus also used from Contains. The HashCode for an edge is only calculated once during Add(). A HashCode for an object should be immutable - we don't want it to change over time - thats why I refactored the Edge class to ensure the Source, Target and Label cannot be changed from the outside.

Comment (Zoe-ms) [inline WCFDataService/Client/System/Data/Services/Client/Binding/BindingGraph.cs]: I see. Thanks, Uffe.

**Comment by LaylaLiu** [comment]
Checked in by commit https://github.com/OData/odata.net/commit/e101af0c030047b5c3c41849f6289272f50b8863

**Pull request closed without merging**